### PR TITLE
Port of AHC RxJava integration to RxJava 2

### PR DESCRIPTION
**Pull request by tobias-bahls**
Straight-up port from [async-http-client-extras-rxjava](https://github.com/AsyncHttpClient/async-http-client/tree/master/extras/rxjava)
to RxJava 2.

Once we know that this implementation is stable, this will be submitted as a pull request to the AHC repo.

The only behavioural change is that AsyncHttpSingle emits onError(AbortedException) instead of onComplete(null) if the underlying AsyncHandler aborted the request. This change was necessary because RxJava2 does not allow emitting nulls.

### CODE REVIEW

File: src/main/java/org/asynchttpclient/extras/rxjava2/AbortedException.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright (c) 2017 AsyncHttpClient Project. All rights reserved.
```
Comment (cberg-zalando) [inline src/main/java/org/asynchttpclient/extras/rxjava2/AbortedException.java]: Should we include that header inside our own library? Perhaps its better to remove that as undertaking is currently licensed using MIT.

File: src/test/java/org/asynchttpclient/extras/rxjava2/single/AsyncHttpSingleTest.java [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+
+import io.reactivex.Single;
+import io.reactivex.exceptions.CompositeException;
+import io.reactivex.observers.TestObserver;
+
+public class AsyncHttpSingleTest {
+
+    @Test(expected = NullPointerException.class )
+    public void testFailsOnNullRequest() {
+        org.asynchttpclient.extras.rxjava2.single.AsyncHttpSingle.create((BoundRequestBuilder) null);
```
Comment (cberg-zalando) [inline src/test/java/org/asynchttpclient/extras/rxjava2/single/AsyncHttpSingleTest.java]: Do we really need the fully qualified import here?

File: src/main/java/org/asynchttpclient/extras/rxjava2/single/AbstractSingleSubscriberBridge.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+import io.reactivex.SingleEmitter;
+import io.reactivex.exceptions.CompositeException;
+import io.reactivex.exceptions.Exceptions;
+
+abstract public class AbstractSingleSubscriberBridge<T> implements AsyncHandler<Void> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(AbstractSingleSubscriberBridge.class);
+
+    protected final SingleEmitter<T> subscriber;
```
Comment (cberg-zalando) [inline src/main/java/org/asynchttpclient/extras/rxjava2/single/AbstractSingleSubscriberBridge.java]: We should name this variable emitter now, I would say.

Comment (cberg-zalando) [inline src/main/java/org/asynchttpclient/extras/rxjava2/single/AbstractSingleSubscriberBridge.java]: And also change child classes and method parameters accordingly.

File: src/test/java/org/asynchttpclient/extras/rxjava2/AsyncHttpObservableTest.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+import static org.asynchttpclient.Dsl.*;
+
+import java.util.concurrent.TimeUnit;
+
+import org.asynchttpclient.AsyncHttpClient;
+import org.junit.Test;
+
+import io.reactivex.Observable;
+
```
Comment (cberg-zalando) [inline src/test/java/org/asynchttpclient/extras/rxjava2/AsyncHttpObservableTest.java]: :frog:

File: src/test/java/org/asynchttpclient/extras/rxjava2/AsyncHttpObservableTest.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    public void testToFlowableError() {
+        try (AsyncHttpClient client = asyncHttpClient()) {
+            AsyncHttpObservable.toObservable(() -> client.prepareGet("http://gatling.io/ttfn"))
+                    .test()
+                    .awaitDone(5, TimeUnit.SECONDS)
+                    .assertValue(resp -> resp.getStatusCode() == 404);
+        } catch (Exception e) {
+            Thread.currentThread().interrupt();
+        }
+
```
Comment (cberg-zalando) [inline src/test/java/org/asynchttpclient/extras/rxjava2/AsyncHttpObservableTest.java]: :frog:

File: src/test/java/org/asynchttpclient/extras/rxjava2/AsyncHttpObservableTest.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+import org.asynchttpclient.AsyncHttpClient;
+import org.junit.Test;
+
+import io.reactivex.Observable;
+
+
+public class AsyncHttpObservableTest {
+
+    @Test
+    public void testToFlowableNoError() {
```
Comment (cberg-zalando) [inline src/test/java/org/asynchttpclient/extras/rxjava2/AsyncHttpObservableTest.java]: s/Flowable/Observable/g

File: src/main/java/org/asynchttpclient/extras/rxjava2/AsyncHttpObservable.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+import io.reactivex.Observable;
+import io.reactivex.subjects.ReplaySubject;
+
+/**
+ * Provide RxJava support for executing requests. Request can be subscribed to and manipulated as needed.
+ *
+ * @see <a href="https://github.com/ReactiveX/RxJava">https://github.com/ReactiveX/RxJava</a>
+ */
+public class AsyncHttpObservable {
```
Comment (cberg-zalando) [inline src/main/java/org/asynchttpclient/extras/rxjava2/AsyncHttpObservable.java]: As we want to support RxJava2, what about providing additionally an `AsyncHttpFlowable` class?

Comment (tobias-bahls) [inline src/main/java/org/asynchttpclient/extras/rxjava2/AsyncHttpObservable.java]: Right now, I don't think there is an use case for that.

**Comment by cberg-zalando** [comment]
:+1:

**Comment by tobias-bahls** [comment]
:+1:

**Pull request merged**